src/tbb/arena.h

```diff
@@ -179,6 +179,82 @@ class atomic_flag {
+            my_state.store(platform_policy, std::memory_order_relaxed);
+        } else {
+            __TBB_ASSERT(lp == tbb::task_arena::leave_policy::fast,
+                         "Was the new value introduced for leave policy?");
+            my_state.store(FAST_LEAVE, std::memory_order_relaxed);
+        }
+    }
+
+    void reset_if_needed() {
+        std::uintptr_t curr = ONE_TIME_FAST_LEAVE;
+        if (my_state.load(std::memory_order_relaxed) == curr) {
+            // Potentially can override decision of the parallel block from future epoch
+            // but it is not a problem because it does not violate the correctness
+            my_state.compare_exchange_strong(curr, DELAYED_LEAVE);
+        }
+    }
+
+    // Indicate start of parallel phase in the state machine
+    void register_parallel_phase() {
+        std::uintptr_t prev = my_state.load(std::memory_order_relaxed);
+        __TBB_ASSERT(prev != std::uintptr_t(-1), "The initial state was not set");
+
+        std::uintptr_t desired{};
+        do {
+            // Need to add a reference for this start of a parallel phase, preserving the leave
+            // policy. Except for the case when one time fast leave was requested at the end of a
+            // previous phase.
+            desired = prev;
+            if (prev == ONE_TIME_FAST_LEAVE) {
+                // State was previously transitioned to "One-time Fast leave", thus with the start
+                // of new parallel phase, it should be transitioned to "Delayed leave"
+                desired = DELAYED_LEAVE;
+            }
+            __TBB_ASSERT(desired + PARALLEL_PHASE > desired, "Overflow detected");
+            desired += PARALLEL_PHASE; // Take into account this start of a parallel phase
+        } while (!my_state.compare_exchange_strong(prev, desired));
+    }
+
+    // Indicate the end of parallel phase in the state machine
+    void unregister_parallel_phase(bool enable_fast_leave) {
+        std::uintptr_t prev = my_state.load(std::memory_order_relaxed);
+        __TBB_ASSERT(prev != std::uintptr_t(-1), "The initial state was not set");
+
+        std::uintptr_t desired{};
+        do {
+            __TBB_ASSERT(prev - PARALLEL_PHASE < prev,
+                         "A call to unregister without its register complement");
+            desired = prev - PARALLEL_PHASE; // Mark the end of this phase in reference counter
+            if (enable_fast_leave && /*it was the last parallel phase*/desired == DELAYED_LEAVE) {
+                desired = ONE_TIME_FAST_LEAVE;
+            }
+        } while (!my_state.compare_exchange_strong(prev, desired));
+    }
+
+    bool is_retention_allowed() {
+        std::uintptr_t curr = my_state.load(std::memory_order_relaxed);
+        __TBB_ASSERT(curr != std::uintptr_t(-1), "The initial state was not set");
+        return curr != FAST_LEAVE && curr != ONE_TIME_FAST_LEAVE;
```

Comment:
Should not it be `curr & (FAST_LEAVE | ONE_TIME_FAST_LEAVE) != 0` instead, to account for the parallel phase counter?

Reply:
The idea of current implementation is that when parallel phase is active then threads are retained regardless of arena default leave policy (as it was said in RFC, scheduler can implement different retention policy). This allows to support this feature on hybrid platforms as well where worker threads are not retained by default, but if the phase is active, they are.

That is why the check is like that. It is used to verify if parallel phase isn't active anymore, meaning threads can fall back to default leave policy (fast or block time) or leave once fast if such option is chosen.